File: libplugin/src/main/java/com/lemon/plugin/inject/tool/asm/AsmInjectTool.java
package com.lemon.plugin.inject.tool.asm;

import com.lemon.plugin.inject.InjectUtils;
import com.lemon.plugin.inject.IInjectTool;

import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

public class AsmInjectTool implements IInjectTool {
    @Override
    public void injectDir(File dir) {

    }

    @Override
    public void injectFile(File dir, File file) {
        try {
            ClassReader classReader = new ClassReader(new FileInputStream(file));
            ClassWriter classWriter = new ClassWriter(classReader,ClassWriter.COMPUTE_MAXS);
            ClassVisitor classVisitor = new ClassVisitor(Opcodes.ASM5,classWriter) {
                @Override
                public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
                    InjectUtils.println("visitMethod() called with: access = [" + access + "], name = [" + name + "], desc = [" + desc + "], signature = [" + signature + "], exceptions = [" + exceptions + "]");
                    return super.visitMethod(access, name, desc, signature, exceptions);
                }

                @Override
                public void visitSource(String source, String debug) {
                    InjectUtils.println("visitSource() called with: source = [" + source + "], debug = [" + debug + "]");
                    super.visitSource(source, debug);
                }
            };
            classReader.accept(classVisitor, ClassReader.EXPAND_FRAMES);
            byte[] code = classWriter.toByteArray();
            FileOutputStream fos = new FileOutputStream(file.getParentFile().getAbsolutePath() + File.separator + file.getName());
            fos.write(code);
            fos.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: libplugin/src/main/groovy/com/creation/plugin/inject/tool/asm/AsmInjectTool.kt
package com.creation.plugin.inject.tool.asm

import com.creation.plugin.inject.InjectUtils
import com.creation.plugin.inject.IInjectTool

import org.objectweb.asm.ClassReader
import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.ClassWriter
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes

import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

class AsmInjectTool : IInjectTool {
    override fun injectDir(dir: File) {

    }

    override fun injectFile(dir: File, file: File) {
        try {
            val classReader = ClassReader(FileInputStream(file))
            val classWriter = ClassWriter(classReader, ClassWriter.COMPUTE_MAXS)
            val classVisitor = object : ClassVisitor(Opcodes.ASM5, classWriter) {
                override fun visitMethod(access: Int, name: String, desc: String, signature: String, exceptions: Array<String>): MethodVisitor {
                    InjectUtils.println("visitMethod() called with: access = [$access], name = [$name], desc = [$desc], signature = [$signature], exceptions = [$exceptions]")
                    return super.visitMethod(access, name, desc, signature, exceptions)
                }

                override fun visitSource(source: String, debug: String) {
                    InjectUtils.println("visitSource() called with: source = [$source], debug = [$debug]")
                    super.visitSource(source, debug)
                }
            }
            classReader.accept(classVisitor, ClassReader.EXPAND_FRAMES)
            val code = classWriter.toByteArray()
            val fos = FileOutputStream(file.parentFile.absolutePath + File.separator + file.name)
            fos.write(code)
            fos.close()
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

    }
}
